### ISSUE TITLE
[Accessibility] Add audio cues for indentation levels

### ISSUE THREAD

**Issue by jooyoungseo** [microsoft/vscode#147386]
CC @isidorn @hediet

Indentation level (depthness) is very critical in understanding code structures, and sometimes it's part of syntax rules, such as Python. Although NVDA has a cool feature to represent indentation level using  beep tones (a deeper indentation corresponds to a higher beep tone), not every screen reader has such feature. For example, JAWS, Narrator, and VoiceOver do not have that configuration.

I strongly suggest VSCode implement this feature via Audio Cues so that screen-reader users can universally benefit from this feature regardless of their screen reader or OS types. NVDA users can use either their screen reader feature or VSCode one, depending on their needs.

**Comment by isidorn** [comment]
@jooyoungseo thanks for the suggestion. Ideally I would prefer if Screen Readers supported this on their side.
@joanmarie what do you think? Does Orca have a way to read indentation like NVDA?

If there is a reason why screen reader can not do this then we can think about adding this on the VS Code side.

**Comment by joanmarie** [comment]
@isidorn, you have to explicitly enable presentation of indentation and justification, but then yes, Orca speaks it. We can ask @jvesouza if there are any bugs or missing features in this area.

**Comment by jvesouza** [comment]
When I navigate with the arrows in the VSCode editor, orca announces the amount of spaces or tabs contained at the beginning of the line. This lets me get a good idea of ​​the indentation of what I'm coding.
One feature that might be useful was being able to quickly jump to the next change in indentation.
Let's say we're exploring code that is always starting at column 4. You might want to quickly jump to the next or previous line that doesn't start at column 4.

**Comment by jooyoungseo** [comment]
@isidorn -- I would like to bring your attention to the difference between "spoken output" and "audio cues".

Currently, the way that JAWS and ORCA deliver the indentation is done via "spoken output," such as "four spaces," "eight spaces" etc. This kind of information representation is processed via verbal channel in our brains.

On the other hand, NVDA uses audio cues (AKA, earcons) using different pitch tones that are processed via non-verbal channel in our brains.

Since blind programmers already rely on verbal channel to understand code syntax, using additional verbal representation for indentation levels is not quite cogntively effective.

I beg to differ with the idea that this feature has to be taken care of by each screen reader  because there is no guarantee FreedomScientific (for JAWS), Apple (VoiceOver), Microsoft (for Narrator), and other screen reader makers would implement this feature sooner or later. As far as I know, many JAWS and voiceOver users already requested this feature to each company; unfortunately, however, it has never been implemented for a long time already.

That's why [Indent Report](https://marketplace.visualstudio.com/items?itemName=ogomez92.indent-report) was developed; however, this extension is no longer working, nor reliable.

Given the importance of indentation in the context of programming and the current VSCode's position as one of the most accessible code editors, I still strongly suggest that your team seriously consider adding this option so that screen reader users can better customize their coding environments in a more accessible manner.

**Comment by jooyoungseo** [comment]
@jvesouza, take a look at [vscode-indent-nav](https://github.com/mltony/vscode-indent-nav) that will do what you want.

**Comment by jhomme** [comment]
JAWS has such a capability. I think this should be an extension and wish I had the chops to code it.

**Comment by jooyoungseo** [comment]
@jhomme -- Be careful about any  possibly misleading information here please.

Note that JAWS "Speech and Sound Schemes" and "Dictionary Manager" do not allow regex patterns, and currently there is no direct way to implement audio-based indentation unless you write a JAWS script.

In other words, what's possible with the JAWS "Speech and Sound Schemes" and "Dictionary Manager" is a fixed pattern match, not an intelligent indent detection. this fact was directly confirmed by one of the JAWS developers. He admitted that JAWS does not have such a way to customize audio-indent just like NVDA.

**Comment by isidorn** [comment]
Thanks for a great discussion.

I think for this to work the sound would need to be programmable - since we want to achieve higher tones based on indentation. And that is not trivial, and something that the web sound API might not support @hediet 
We can not just hard code some sounds, since we do not know how deep the indentation might go.

@jooyoungseo I was not aware of that extension, I just created an issue for @jhomme to publish it to the marketplace so more people can find it easily.

**Comment by hediet** [comment]
> and something that the web sound API might not support

Actually, this is possible, but not trivial.
